vm-allocator/src/address.rs

```diff
@@ -0,0 +1,365 @@
+            end,
+            alignment,
+            ranges: BTreeMap::new(),
+        };
+
+        // Insert the last address as a zero size range.
+        // This is our end of address space marker.
+        allocator.ranges.insert(base.checked_add(size)?, 0);
+
+        Some(allocator)
+    }
+
+    fn align_address(&self, address: GuestAddress) -> GuestAddress {
+        let align_adjust = if address.raw_value() % self.alignment != 0 {
+            self.alignment - (address.raw_value() % self.alignment)
+        } else {
+            0
+        };
+
+        address.unchecked_add(align_adjust)
+    }
+
+    fn available_range(
+        &self,
+        req_address: GuestAddress,
+        req_size: GuestUsize,
+    ) -> Result<GuestAddress> {
+        let aligned_address = self.align_address(req_address);
+
+        // The requested address should be aligned.
+        if aligned_address != req_address {
+            return Err(Error::UnalignedAddress);
+        }
+
+        // The aligned address should be within the address space range.
+        if aligned_address >= self.end || aligned_address <= self.base {
+            return Err(Error::Overflow);
+        }
+
+        let mut prev_end_address = self.base;
+        for (address, size) in self.ranges.iter() {
+            if aligned_address <= *address {
+                // Do we overlap with the previous range?
+                if prev_end_address > aligned_address {
+                    return Err(Error::Overlap);
+                }
+
+                // Do we have enough space?
+                if address
+                    .unchecked_sub(aligned_address.raw_value())
+                    .raw_value()
+                    < req_size
+                {
+                    return Err(Error::Overlap);
+                }
+
+                return Ok(aligned_address);
+            }
+
+            prev_end_address = address.unchecked_add(*size);
+        }
+
+        // We have not found a range that starts after the requested address,
+        // despite having a marker at the end of our range.
+        Err(Error::Overflow)
+    }
+
+    fn first_available_range(&self, req_size: GuestUsize) -> Result<GuestAddress> {
+        let mut prev_end_address = self.base;
+
+        for (address, size) in self.ranges.iter() {
+            // If we have enough space between this range and the previous one,
+            // we return the start of this range minus the requested size.
+            // As each new range is allocated at the end of the available address space,
+            // we will tend to always allocate new ranges there as well. In other words,
+            // ranges accumulate at the end of the address space.
+            if address
+                .unchecked_sub(self.align_address(prev_end_address).raw_value())
+                .raw_value()
+                >= req_size
+            {
+                return Ok(self.align_address(address.unchecked_sub(req_size + self.alignment)));
```

Comment:
You'r right. Also this would have hole when req_size is aligned. What about this one?
```
       let req_align = self.align_address(GuestAddress(req_size))?;
       let addr = self.align_address(address.unchecked_sub(req_align.raw_value()))?;
```